File: odoo/addons/weselaja_payment/services/weselaja_client.py
import base64
import hashlib
import hmac
import json
import uuid
from datetime import datetime, timezone

import requests


PAYMENT_URI = "/v1/payins"


class WeselAjaClient:

    def __init__(self, provider):
        self.provider = provider

        self.api_url = (
            provider.weselaja_api_url or ""
        ).rstrip("/")

        self.api_key = (
            provider.weselaja_api_key or ""
        )

        self.secret_key = (
            provider.weselaja_secret_key or ""
        )

    def is_configured(self):
        return bool(
            self.api_url
            and self.api_key
            and self.secret_key
        )

    @staticmethod
    def generate_signature(
        method,
        uri,
        timestamp,
        body,
        secret_key,
    ):
        payload = (
            f"{method}\n"
            f"{uri}\n"
            f"{timestamp}\n"
            f"{body}"
        )

        digest = hmac.new(
            secret_key.encode("utf-8"),
            payload.encode("utf-8"),
            hashlib.sha256,
        ).digest()

        return base64.b64encode(
            digest
        ).decode("utf-8")

    def create_payment(self, arguments):

        if not self.is_configured():
            return {
                "success": False,
                "paymentId": None,
                "initiatedAmount": None,
                "paymentAmount": None,
                "feeAmount": None,
                "currency": None,
                "paymentMethod": None,
                "paymentChannel": None,
                "paymentCode": None,
                "paymentCodeType": None,
                "referenceCode": None,
                "customerReference": None,
                "customerName": None,
                "status": None,
                "createdTime": None,
                "updatedTime": None,
                "expirationTime": None,
                "description": None,
                "callbackUrl": None,
                "redirectUrl": None,
                "payerAccountName": None,
                "payerAccountNumber": None,
                "payerPaymentChannel": None,
                "metadata": {},
                "error":
                    "WeselAja provider is not configured.",
            }

        if not isinstance(arguments, dict):
            return {
                "success": False,
                "error":
                    "arguments must be an object.",
            }

        required_fields = [
            "amount",
            "currency",
            "paymentMethod",
            "paymentChannel",
            "referenceCode",
            "customerReference",
            "customerName",
            "callbackUrl",
            "redirectUrl",
        ]

        for field in required_fields:

            value = arguments.get(field)

            if value is None:
                return {
                    "success": False,
                    "error":
                        f"{field} is required.",
                }

            if (
                isinstance(value, str)
                and not value.strip()
            ):
                return {
                    "success": False,
                    "error":
                        f"{field} is required.",
                }

        try:
            amount = int(arguments["amount"])
        except (
            TypeError,
            ValueError,
        ):
            return {
                "success": False,
                "error":
                    "amount must be a number.",
            }

        if amount <= 0:
            return {
                "success": False,
                "error":
                    "Payment amount must be greater than zero.",
            }

        body_data = {
            "initiatedAmount": amount,
            "currency":
                str(arguments["currency"]),
            "paymentMethod":
                str(arguments["paymentMethod"]),
            "paymentChannel":
                str(arguments["paymentChannel"]),
            "referenceCode":
                str(arguments["referenceCode"]),
            "customerReference":
                arguments["customerReference"],
            "customerName":
                str(arguments["customerName"]),
            "callbackUrl":
                str(arguments["callbackUrl"]),
            "redirectUrl":
                str(arguments["redirectUrl"]),
        }

        customer_phone_number = arguments.get(
                "customerPhoneNumber"
            )

        if (
            customer_phone_number is not None
            and str(
                customer_phone_number
            ).strip()
        ):
            body_data[
                "customerPhoneNumber"
            ] = str(
                customer_phone_number
            )

        description = arguments.get("description")

        if (
            description is not None
            and str(description).strip()
        ):
            body_data[
                "description"
            ] = str(description)

        body = json.dumps(
                body_data,
                separators=(",", ":"),
                ensure_ascii=False,
            )

        timestamp = (
                datetime.now(
                    timezone.utc
                )
                .isoformat(
                    timespec="milliseconds"
                )
                .replace(
                    "+00:00",
                    "Z",
                )
            )

        signature = self.generate_signature(
                "POST",
                PAYMENT_URI,
                timestamp,
                body,
                self.secret_key,
            )

        idempotency_key ="odoo-" + uuid.uuid4().hex

        url = self.api_url + PAYMENT_URI

        headers = {
            "Content-Type":
                "application/json",

            "Accept":
                "application/json",

            "Xenith-Api-Key":
                self.api_key,

            "Xenith-Request-Timestamp":
                timestamp,

            "Xenith-Request-Signature":
                signature,

            "X-Idempotency-Key":
                idempotency_key,
        }

        try:
            response = requests.post(
                    url,
                    data=body.encode("utf-8"),
                    headers=headers,
                    timeout=30,
                )

        except requests.RequestException as exc:

            return {
                "success": False,
                "error":
                    "Unable to connect to WeselAja: "
                    f"{exc}",
            }

        try:
            response_data = response.json()

        except ValueError:

            return {
                "success": False,
                "error":
                    "Invalid JSON response from WeselAja. "
                    f"HTTP {response.status_code}: "
                    f"{response.text}",
            }

        if not isinstance(
            response_data,
            dict,
        ):
            return {
                "success": False,
                "error":
                    "Unexpected WeselAja response format.",
            }

        if (
            response.status_code < 200
            or response.status_code >= 300
        ):
            api_error = (
                response_data.get("message")
                or response_data.get("error")
                or response_data.get("code")
                or response.text
            )

            if isinstance(
                api_error,
                dict,
            ):
                api_error = json.dumps(
                        api_error,
                        ensure_ascii=False,
                    )

            return {
                "success": False,
                "error":
                    str(api_error),
            }

        return {
            "success": True,

            "paymentId":
                response_data.get("id"),

            "initiatedAmount":
                response_data.get(
                    "initiatedAmount"
                ),

            "paymentAmount":
                response_data.get(
                    "paymentAmount"
                ),

            "feeAmount":
                response_data.get(
                    "feeAmount"
                ),

            "currency":
                response_data.get(
                    "currency"
                ),

            "paymentMethod":
                response_data.get(
                    "paymentMethod"
                ),

            "paymentChannel":
                response_data.get(
                    "paymentChannel"
                ),

            "paymentCode":
                response_data.get(
                    "paymentCode"
                ) or response_data.get(
                    "paymentUrl"
                ),

            "paymentCodeType":
                response_data.get(
                    "paymentCodeType"
                ),

            "referenceCode":
                response_data.get(
                    "referenceCode"
                ),

            "customerReference":
                response_data.get(
                    "customerReference"
                ),

            "customerName":
                response_data.get(
                    "customerName"
                ),

            "status":
                response_data.get(
                    "status"
                ),

            "createdTime":
                response_data.get(
                    "createdTime"
                ),

            "updatedTime":
                response_data.get(
                    "updatedTime"
                ),

            "expirationTime":
                response_data.get(
                    "expirationTime"
                ),

            "description":
                response_data.get(
                    "description"
                ),

            "callbackUrl":
                response_data.get(
                    "callbackUrl"
                ),

            "redirectUrl":
                response_data.get(
                    "redirectUrl"
                ),

            "payerAccountName":
                response_data.get(
                    "payerAccountName"
                ),

            "payerAccountNumber":
                response_data.get(
                    "payerAccountNumber"
                ),

            "payerPaymentChannel":
                response_data.get(
                    "payerPaymentChannel"
                ),

            "metadata":
                response_data.get(
                    "metadata"
                )
                or {},

            "error":
                None,
        }

File: csharp/WeselAja/Services/WeselAjaClient.cs
using System.Diagnostics;
using System.Globalization;
using System.Net;
using System.Net.Http.Headers;
using System.Security.Cryptography;
using System.Text;
using System.Text.Encodings.Web;
using System.Text.Json;
using WeselAjaMcp.Configuration;
using WeselAjaMcp.Models;
using WeselAjaMcp.Security;

namespace WeselAjaMcp.Services;

public class WeselAjaClient : IWeselAjaClient
{
    private readonly HttpClient _httpClient;
    private readonly WeselAjaOptions _options;
    private readonly SignatureGenerator _signatureGenerator;

    private const string PaymentUri = "/v1/payins";
    private const string UserAgent = "Python-urllib/3.13";

    public WeselAjaClient(
        HttpClient httpClient,
        WeselAjaOptions options,
        SignatureGenerator signatureGenerator)
    {
        _httpClient = httpClient;
        _options = options;
        _signatureGenerator = signatureGenerator;
    }

    public async Task<CreatePaymentResponse> CreatePaymentAsync(
        CreatePaymentRequest request,
        CancellationToken cancellationToken = default)
    {
        if (!_options.Enabled)
        {
            return ErrorResponse(
                "XenithPay is disabled.");
        }

        var apiUrl =
            _options.ApiUrl?.Trim()
            ?? string.Empty;

        var apiKey =
            _options.ApiKey?.Trim()
            ?? string.Empty;

        var secretKey =
            _options.SecretKey?.Trim()
            ?? string.Empty;

        if (string.IsNullOrWhiteSpace(apiUrl))
        {
            return ErrorResponse(
                "WeselAja API URL is not configured.");
        }

        if (string.IsNullOrWhiteSpace(apiKey))
        {
            return ErrorResponse(
                "XenithPay API key is not configured.");
        }

        if (string.IsNullOrWhiteSpace(secretKey))
        {
            return ErrorResponse(
                "XenithPay secret key is not configured.");
        }

        if (request.InitiatedAmount <= 0)
        {
            return ErrorResponse(
                "Payment amount must be greater than zero.");
        }

        if (string.IsNullOrWhiteSpace(request.Currency))
        {
            return ErrorResponse(
                "Currency is required.");
        }

        if (string.IsNullOrWhiteSpace(request.PaymentMethod))
        {
            return ErrorResponse(
                "Payment method is required.");
        }

        if (string.IsNullOrWhiteSpace(request.PaymentChannel))
        {
            return ErrorResponse(
                "Payment channel is required.");
        }

        if (string.IsNullOrWhiteSpace(request.ReferenceCode))
        {
            return ErrorResponse(
                "Reference code is required.");
        }

        if (string.IsNullOrWhiteSpace(request.CustomerReference))
        {
            return ErrorResponse(
                "Customer reference is required.");
        }

        if (string.IsNullOrWhiteSpace(request.CustomerName))
        {
            return ErrorResponse(
                "Customer name is required.");
        }

        if (string.IsNullOrWhiteSpace(request.CallbackUrl))
        {
            return ErrorResponse(
                "Callback URL is required.");
        }

        if (string.IsNullOrWhiteSpace(request.RedirectUrl))
        {
            return ErrorResponse(
                "Redirect URL is required.");
        }

        if (!IsValidReference(request.ReferenceCode))
        {
            return ErrorResponse(
                "Reference code may only contain letters, numbers, '-' and '_'.");
        }

        if (!IsValidReference(request.CustomerReference))
        {
            return ErrorResponse(
                "Customer reference may only contain letters, numbers, '-' and '_'.");
        }

        var payload =
            new Dictionary<string, object?>
            {
                ["initiatedAmount"] =
                    request.InitiatedAmount,

                ["currency"] =
                    request.Currency,

                ["paymentMethod"] =
                    request.PaymentMethod,

                ["paymentChannel"] =
                    request.PaymentChannel,

                ["referenceCode"] =
                    request.ReferenceCode,

                ["customerReference"] =
                    request.CustomerReference,

                ["customerName"] =
                    request.CustomerName,

                ["callbackUrl"] =
                    request.CallbackUrl,

                ["redirectUrl"] =
                    request.RedirectUrl
            };

        if (
            !string.IsNullOrWhiteSpace(
                request.CustomerPhoneNumber))
        {
            payload[
                "customerPhoneNumber"
            ] =
                request.CustomerPhoneNumber;
        }

        if (
            !string.IsNullOrWhiteSpace(
                request.Description))
        {
            payload[
                "description"
            ] =
                request.Description;
        }

        var jsonOptions =
            new JsonSerializerOptions
            {
                WriteIndented = false,

                Encoder =
                    JavaScriptEncoder.UnsafeRelaxedJsonEscaping,

                DefaultIgnoreCondition =
                    System.Text.Json.Serialization
                        .JsonIgnoreCondition
                        .WhenWritingNull
            };

        var body =
            JsonSerializer.Serialize(
                payload,
                jsonOptions);

        var timestamp =
            DateTimeOffset.UtcNow.ToString(
                "yyyy-MM-dd'T'HH:mm:ss.fff'Z'",
                CultureInfo.InvariantCulture);

        var signature =
            _signatureGenerator.Generate(
                "POST",
                PaymentUri,
                timestamp,
                body,
                secretKey);

        var idempotencyKey =
            "csharp-" +
            Guid.NewGuid().ToString("N");

        var requestUrl =
            apiUrl.TrimEnd('/') +
            PaymentUri;

        using var httpRequest =
            new HttpRequestMessage(
                HttpMethod.Post,
                requestUrl)
            {
                Version =
                    HttpVersion.Version11,

                VersionPolicy =
                    HttpVersionPolicy.RequestVersionExact
            };

        var bodyBytes =
            Encoding.UTF8.GetBytes(
                body);

        var content =
            new ByteArrayContent(
                bodyBytes);

        content.Headers.ContentType =
            new MediaTypeHeaderValue(
                "application/json");

        httpRequest.Content =
            content;

        httpRequest.Headers.Accept.Clear();

        httpRequest.Headers.Accept.Add(
            new MediaTypeWithQualityHeaderValue(
                "application/json"));

        httpRequest.Headers.UserAgent.Clear();

        httpRequest.Headers.UserAgent.ParseAdd(
            UserAgent);

        httpRequest.Headers.AcceptEncoding.Clear();

        httpRequest.Headers.AcceptEncoding.Add(
            new StringWithQualityHeaderValue(
                "identity"));

        httpRequest.Headers.ConnectionClose =
            true;

        httpRequest.Headers.ExpectContinue =
            false;

        httpRequest.Headers.Add(
            "Xenith-Api-Key",
            apiKey);

        httpRequest.Headers.Add(
            "Xenith-Request-Timestamp",
            timestamp);

        httpRequest.Headers.Add(
            "Xenith-Request-Signature",
            signature);

        httpRequest.Headers.Add(
            "X-Idempotency-Key",
            idempotencyKey);

        try
        {
            using var response = await _httpClient.SendAsync(
                httpRequest,
                HttpCompletionOption.ResponseContentRead,
                cancellationToken);

            var responseBody = await response.Content.ReadAsStringAsync(
                cancellationToken);

            if (string.IsNullOrWhiteSpace(responseBody))
            {
                return ErrorResponse(
                    $"Empty response from WeselAja. HTTP {(int)response.StatusCode} {response.StatusCode}");
            }

            if (!response.IsSuccessStatusCode)
            {
                return ErrorResponse(
                    ExtractErrorFromResponse(
                        responseBody,
                        response.StatusCode));
            }

            try
            {
                using var document = JsonDocument.Parse(responseBody);

                var root = document.RootElement;

                if (root.ValueKind != JsonValueKind.Object)
                {
                    return ErrorResponse(
                        "WeselAja returned an invalid payment response.");
                }

                if (
                    root.TryGetProperty("success", out var successProperty) &&
                    successProperty.ValueKind == JsonValueKind.False)
                {
                    return ErrorResponse(
                        ExtractErrorFromJson(
                            root,
                            responseBody));
                }

                if (
                    root.TryGetProperty("data", out var dataProperty) &&
                    dataProperty.ValueKind == JsonValueKind.Object)
                {
                    return MapResponse(
                        dataProperty,
                        request);
                }

                var paymentId = GetString(
                    root,
                    "id",
                    "session");

                if (string.IsNullOrWhiteSpace(paymentId))
                {
                    return ErrorResponse(
                        ExtractErrorFromJson(
                            root,
                            responseBody));
                }

                return MapResponse(
                    root,
                    request);
            }
            catch (JsonException exception)
            {
                return ErrorResponse(
                    $"Invalid JSON response from WeselAja. " +
                    $"HTTP {(int)response.StatusCode}: {exception.Message}");
            }
        }
        catch (HttpRequestException exception)
        {
            return ErrorResponse(
                $"HTTP request to WeselAja failed: {exception.Message}");
        }
        catch (TaskCanceledException)
        {
            return ErrorResponse(
                "WeselAja HTTP request timed out or was cancelled.");
        }
    }

    private static string Sha256(
        string value)
    {
        return Sha256(
            Encoding.UTF8.GetBytes(
                value));
    }

    private static string Sha256(
        byte[] value)
    {
        return Convert.ToHexString(
            SHA256.HashData(
                value))
            .ToLowerInvariant();
    }

    private static string MaskSecretHeader(
        string value)
    {
        if (
            string.IsNullOrWhiteSpace(
                value))
        {
            return "***";
        }

        if (
            value.Length <= 8)
        {
            return "***";
        }

        return
            value[..4] +
            "***" +
            value[^4..];
    }

    private static CreatePaymentResponse MapResponse(
        JsonElement data,
        CreatePaymentRequest request)
    {
        return new CreatePaymentResponse
        {
            Id =
                GetString(
                    data,
                    "id",
                    "session"),

            InitiatedAmount =
                GetDecimal(
                    data,
                    "initiatedAmount",
                    "amount")
                ??
                request.InitiatedAmount,

            PaymentAmount =
                GetDecimal(
                    data,
                    "paymentAmount"),

            FeeAmount =
                GetDecimal(
                    data,
                    "feeAmount"),

            Currency =
                GetString(
                    data,
                    "currency")
                ??
                request.Currency,

            PaymentMethod =
                GetString(
                    data,
                    "paymentMethod")
                ??
                request.PaymentMethod,

            PaymentChannel =
                GetString(
                    data,
                    "paymentChannel")
                ??
                request.PaymentChannel,

            PaymentCode =
                GetString(
                    data,
                    "paymentCode",
                    "paymentUrl"),

            PaymentCodeType =
                GetString(
                    data,
                    "paymentCodeType"),

            ReferenceCode =
                GetString(
                    data,
                    "referenceCode")
                ??
                request.ReferenceCode,

            CustomerReference =
                GetString(
                    data,
                    "customerReference")
                ??
                request.CustomerReference,

            CustomerName =
                GetString(
                    data,
                    "customerName")
                ??
                request.CustomerName,

            Status =
                GetString(
                    data,
                    "status"),

            CreatedTime =
                GetString(
                    data,
                    "createdTime",
                    "createdAt"),

            UpdatedTime =
                GetString(
                    data,
                    "updatedTime",
                    "updatedAt"),

            ExpirationTime =
                GetString(
                    data,
                    "expirationTime",
                    "expiresAt"),

            Description =
                GetString(
                    data,
                    "description")
                ??
                request.Description,

            CallbackUrl =
                GetString(
                    data,
                    "callbackUrl")
                ??
                request.CallbackUrl,

            RedirectUrl =
                GetString(
                    data,
                    "redirectUrl",
                    "redirect_url")
                ??
                request.RedirectUrl,

            Error =
                null
        };
    }

    private static bool IsValidReference(
        string value)
    {
        foreach (
            var character in value)
        {
            if (
                char.IsLetterOrDigit(
                    character)
                ||
                character == '-'
                ||
                character == '_')
            {
                continue;
            }

            return false;
        }

        return true;
    }

    private static string ExtractErrorFromResponse(
        string responseBody,
        HttpStatusCode statusCode)
    {
        try
        {
            using var document =
                JsonDocument.Parse(
                    responseBody);

            return ExtractErrorFromJson(
                document.RootElement,
                responseBody);
        }
        catch (
            JsonException)
        {
            return
                $"WeselAja returned HTTP " +
                $"{(int)statusCode} {statusCode}: " +
                responseBody;
        }
    }

    private static string ExtractErrorFromJson(
        JsonElement root,
        string rawResponse)
    {
        if (
            root.TryGetProperty(
                "message",
                out var messageProperty)
            &&
            messageProperty.ValueKind ==
            JsonValueKind.String)
        {
            return
                messageProperty.GetString()
                ??
                rawResponse;
        }

        if (
            root.TryGetProperty(
                "error",
                out var errorProperty)
            &&
            errorProperty.ValueKind ==
            JsonValueKind.String)
        {
            return
                errorProperty.GetString()
                ??
                rawResponse;
        }

        if (
            root.TryGetProperty(
                "errors",
                out var errorsProperty))
        {
            return
                errorsProperty.ToString();
        }

        if (
            root.TryGetProperty(
                "code",
                out var codeProperty)
            &&
            codeProperty.ValueKind ==
            JsonValueKind.String)
        {
            return
                codeProperty.GetString()
                ??
                rawResponse;
        }

        return rawResponse;
    }

    private static string? GetString(
        JsonElement element,
        params string[] propertyNames)
    {
        foreach (
            var propertyName in propertyNames)
        {
            if (
                !element.TryGetProperty(
                    propertyName,
                    out var property))
            {
                continue;
            }

            return property.ValueKind switch
            {
                JsonValueKind.String =>
                    property.GetString(),

                JsonValueKind.Number =>
                    property.ToString(),

                JsonValueKind.True =>
                    "true",

                JsonValueKind.False =>
                    "false",

                JsonValueKind.Null =>
                    null,

                _ =>
                    property.ToString()
            };
        }

        return null;
    }

    private static decimal? GetDecimal(
        JsonElement element,
        params string[] propertyNames)
    {
        foreach (
            var propertyName in propertyNames)
        {
            if (
                !element.TryGetProperty(
                    propertyName,
                    out var property))
            {
                continue;
            }

            if (
                property.ValueKind ==
                JsonValueKind.Number
                &&
                property.TryGetDecimal(
                    out var number))
            {
                return number;
            }

            if (
                property.ValueKind ==
                JsonValueKind.String
                &&
                decimal.TryParse(
                    property.GetString(),
                    NumberStyles.Number,
                    CultureInfo.InvariantCulture,
                    out var parsed))
            {
                return parsed;
            }
        }

        return null;
    }

    private static CreatePaymentResponse ErrorResponse(
        string message)
    {
        return new CreatePaymentResponse
        {
            Error = message
        };
    }
}